src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,205 @@
+            add(value);
+        } else if (index == 0) {
+            addToHead(value);
+        } else {
+            Node<T> previous = findNodeByIndex(index);
+            if (previous == null) {
+                addToTail(value);
+            } else {
+                insert(value, previous);
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isValidIndexForSetAndGet(index);
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isValidIndexForSetAndGet(index);
+        Node<T> foundNode = findNodeByIndex(index);
+        T oldValue = foundNode.value;
+        foundNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> deletedElement = findNodeByIndex(index);
+        if (deletedElement == null) {
+            throw new IndexOutOfBoundsException("wrond index: " + index + ", for empty list");
+        }
+        unlink(deletedElement);
+        size--;
+        return deletedElement.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> deletedElement = findNodeByElement(object);
+        boolean isDeleted = unlink(deletedElement);
+        if (isDeleted) {
+            size--;
+        }
+        return isDeleted;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void insert(T value, Node<T> previous) {
+        Node<T> newNode = new Node(previous.prev, value, previous);
+        previous.prev.next = newNode;
+        previous.prev = newNode;
+        size++;
+    }
+
+    private void addToHead(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private void addToTail(T value) {
+        Node<T> newNode = new Node<>(tail, value,null);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void isValidIndexForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("invalid index: " + index);
+        }
+    }
+
+    private void isValidIndexForSetAndGet(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("invalid index: " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        isValidIndexForAdd(index);
+        if (index == 0) {
+            return head;
+        } else if (index == size - 1) {
+            return tail;
+        }
+        if (index < size / 2) {
+            return startFromHead(index);
```

Comment:
method naming, it is not clear from the method name that it is find and return the node.
the same remark for the method `startFromTail()`